lib/iaas/AzureClient.js

```diff
@@ -0,0 +1,192 @@
+'use strict';
+const _ = require('lodash');
+const Promise = require('bluebird');
+const azureStorage = require('azure-storage');
+const logger = require('../logger');
+const errors = require('../errors');
+const utils = require('../utils');
+const ComputeClient = require('./ComputeClient');
+const CONST = require('../constants');
+const BaseCloudClient = require('./BaseCloudClient');
+const NotFound = errors.NotFound;
+const Unauthorized = errors.Unauthorized;
+const Forbidden = errors.Forbidden;
+
+class AzureClient extends BaseCloudClient {
+  constructor(settings) {
+    super(settings);
+    this.storage = this.constructor.createStorageClient(_
+      .chain(this.settings)
+      .omit('name')
+      .set('provider', this.provider)
+      .value()
+    );
+    this.computeClient = this.constructor.createComputeClient(_
+      .chain(this.settings)
+      .set('provider', this.provider)
+      .value()
+    );
+    Promise.promisifyAll(this.computeClient.snapshots);
+
+    this.storage.on('log::*', function onmessage(message, obj) {
+      const event = _.nth(this.event.split('::'), 1);
+      const level = _.includes([
+        'warn',
+        'info',
+        'verbose',
+        'debug'
+      ], event) ? event : 'debug';
+      if (obj) {
+        logger.log(level, message, obj);
+      } else {
+        logger.log(level, message);
+      }
+    });
+  }
+
+  getContainer(container) {
+    if (arguments.length < 1) {
+      container = this.containerName;
+    }
+    return this.storage
+      .getContainerPropertiesAsync(container)
+      .then(result => result);
+  }
+
+  list(container, options) {
+    if (arguments.length < 2) {
+      options = container;
+      container = this.containerName;
+    }
+    const prefix = options ? options.prefix : null;
+    //TODO Need to update following line with value from 'options' e.g. options.marker
+    let continuationToken = null;
+    //TODO : maxResult is 5000 
+    //https://azure.github.io/azure-storage-node/BlobService.html#listBlobsSegmentedWithPrefix__anchor
+    //If want to fetch more use 'continuationToken' returned in result
+    return this.storage  
+      .listBlobsSegmentedWithPrefixAsync(container, prefix, continuationToken)  
```

Comment:
An implementation similar to `ResourceStream` which internally iterates over all results and promisify the method must be done to handle pagination instead of doing it like this Or look at implemenation of `BackupStore.listFilenames` which invokes itself recursively till all the pages are read.

Reply:
should we fetch all in this this method itself ? ..... In case of BackupStore.listFilenames(), the caller (BackupStore) is doing the accumulation. CloudProviderClient also returning batch-by-batch. Not whole. If we do accumulation here, then it behaves differently than CoudProviderClient. Problem is in BackupStore.listFilenames() we have done package cloud specific code. Now somehow we should harmonise that, in order to support azure library structure. According to me this change is not must as part of initial release as Azure supports by default 5000 and very unlikely we'll be having 5000 backups in a space with backup schedule option once per day (instead of 3 per day which is case in AWS). But I agree it is better to have now only. We can discuss face to face  once and take decision.

Reply:
@subhankarc  Yes, for this we need to change in BackupStore.listFilenames(). But want to do that diligently as we have to test with > 5000 blobs and need to incorporate unit test for that as well. As we are going staging this sprint, my suggestion is to keep a track of this and will do it properly. Also, as I mentioned Azure supports 5000 instead of 1000 which is case for AWS and with frequency once per day instead of thrice which is case for AWS. Please let me know if this is fine.